design/design-doc-e2e-testing.md

```diff
@@ -0,0 +1,809 @@
+for our use case. As we could see, it is not fully declarative. For a project
+the size of Crossplane we need something that is easy to develop, readable and
+maintainable in the long run.
+
+### E2E-framework
+
+[kubernetes-sig/e2e-framework] is a new framework for writing imperative
+end-to-end tests for Kubernetes.
+
+#### Anatomy of a test
+
+Tests are organized in test suites, each test suite being a Go package. The
+function TestMain is used to define package-wide testing steps and configure 
+behavior. The framework exposes the Environment interface, which is used to
+interact with the cluster.
+
+Furthermore, the environment is set up by defining steps that are executed in
+different phases of the test, like Setup, Finish, BeforeEach, kinda similar to
+Ginkgo.
+
+TestMain:
+
+```go
+var testenv *env.Environment
+
+func TestMain(m *testing.M) {
+	testenv = env.NewInClusterConfig()
+
+	testenv.Setup(
+		setupSchema,
+		setupDummyProvider,
+	)
+
+	testenv.Finish(
+		teardownDummyProvider,
+	)
+
+	// launch package tests
+	os.Exit(testenv.Run(m))
+}
+```
+
+The tests themselves are organized in Features for which steps (Setup, Assess,
+Teardown) need to be defined.
+
+Test example:
+
+```go
+func TestFlow(t *testing.T) {
+
+	f := features.New("create claim flow").
+		WithLabel("type", "flow-claim").
+		WithSetup("install the XRD", setupXRD).
+		WithTeardown("teardown the XRD", teardownXRD).
+		WithSetup("install the Composition", setupComposition).
+		WithTeardown("teardown the Composition", teardownComposition).
+		WithSetup("xrd is established and offered", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+			xrd := genXRD()
+			err := wait.For(conditions.New(cfg.Client().Resources()).ResourceMatch(xrd, func(object k8s.Object) bool {
+				o := object.(*extv1.CompositeResourceDefinition)
+				return o.Status.GetCondition(extv1.TypeEstablished).Status == corev1.ConditionTrue &&
+					o.Status.GetCondition(extv1.TypeOffered).Status == corev1.ConditionTrue
+			}), wait.WithTimeout(time.Minute*1))
+			if err != nil {
+				t.Fatalf("failed to wait for XRD to be established and offered: %v", err)
+			}
+			return ctx
+		}).
+		Setup(setupClaim).
+		Teardown(teardownClaim).
+		Assess("claim", func(ctx context.Context, t *testing.T, cfg *envconf.Config) context.Context {
+			claim, err := getClaim()
+			if err != nil {
+				t.Fatalf("failed to get claim: %v", err)
+			}
+			err = wait.For(conditions.New(cfg.Client().Resources()).ResourceMatch(claim, func(object k8s.Object) bool {
+				claimObj := composed.Unstructured{Unstructured: *claim}
+				isReady := claimObj.GetCondition(xpv1.TypeReady)
+				if isReady.Status != corev1.ConditionTrue {
+					t.Logf("claim %q is not yet Ready", claim.GetName())
+					return false
+				}
+				return isReady.Status == corev1.ConditionTrue
+			}), wait.WithTimeout(time.Minute*1))
+			if err != nil {
+				t.Fatalf("failed to wait for claim to be Ready: %v", err)
+			}
+			return ctx
```

Comment:
As someone who had a good amount of exposure to the Crossplane codebase and Golang in general, I feel less comfortable with maintaining or extending a test case looking like this compared to what we see with Cucumber above. Yes, you can make it more readable with some refactoring and adding some comments etc. (which is hard to keep quality under control, especially with the test code), but it would never reach that level of readability.

While looking for some other real-life examples of Cucumber, I came across [this test definition](https://github.com/awslabs/aws-cloudformation-controller-for-flux/blob/main/internal/integtests/features/cfn_controller.feature), imagining I need to implement a feature there, I felt very comfortable with extending it.

Reply:
@turkenh Isn't Cucumber pretty similar though if you actually have to add test logic?

Adding a feature seems fairly straightforward (though I worry that in reality you need to understand the regex that binds each to actual test logic). The feature appears to essentially be specifying text fixtures with a bit of flow control though - you still need to write a bunch of Go code to actually do something with the test fixture.

https://github.com/awslabs/aws-cloudformation-controller-for-flux/blob/0dd8444/internal/integtests/cfn_controller_test_scenario.go#L210

I would argue for example the above code is pretty similar to what we see here. It just seems a lot less like writing a Go `testing` test and more like writing arbitrary Go code.

Reply:
I was under the impression that we would write those binding once and rarely need to extend them. This is why I was looking for [the possibility of defining](https://github.com/crossplane/crossplane/pull/4101#discussion_r1206341582) `claim becomes synchronized and ready` as `<resource> becomes <conditions>...` instead.

Reply:
@turkenh @pedjak I spent some time experimenting with e2e-framework in https://github.com/crossplane/crossplane/pull/4187. After doing that, I'm confident we're not making a fair comparison here.

I believe regardless of which of the frameworks we choose it's possible to:

1. Have declarative test fixtures - i.e. YAML manifests as input.
2. Write reusable Go functions to validate that Crossplane does what it should when YAML manifests are applied.

I'm not entirely sure whether @turkenh is comparing the code he's seeing here to the Gherkin language, or to the Go code that powers the Gherkin language.

It does not make sense to compare this code to Gherkin, because Gherkin also need you to write something like this code to function. Gherkin is largely providing and organizing test fixtures that are applied and assessed by Go code (i.e. `godog`).

It does make sense to compare this code to the `godog` code. I believe to the extent that that `godog` code appears simpler, it's because it's largely shelling out to tools like `kubectl` and `jq` to implement the "test logic". I don't think we should do that, but we _could_ just as easily do it with any of the frameworks assessed here.